src/providers/hana/qgshanaexpressioncompiler.cpp

```diff
@@ -0,0 +1,229 @@
+{
+  ok = true;
+  return QgsHanaUtils::quotedValue( value );
+}
+
+static const QMap<QString, QString> FUNCTION_NAMES_SQL_FUNCTIONS_MAP
+{
+  // mathematical functions
+  { "sqrt", "SQRT" },
+  { "sign", "SIGN" },
+  { "abs", "ABS" },
+  { "cos", "COS" },
+  { "sin", "SIN" },
+  { "tan", "TAN" },
+  { "acos", "ACOS" },
+  { "asin", "ASIN" },
+  { "atan", "ATAN" },
+  { "atan2", "ATAN2" },
+  { "exp", "EXP" },
+  { "ln", "LN" },
+  { "log", "LOG" },
+  { "round", "ROUND" },
+  { "floor", "FLOOR" },
+  { "ceil", "CEIL" },
+  // geometry functions
+  { "geom_from_wkt", "ST_GeomFromWKT" },
+  // string functions
+  { "char", "CHAR" },
+  { "lower", "LOWER" },
+  { "upper", "UPPER" },
+  { "trim", "TRIM" },
+  // other helper functions
+  { "coalesce", "COALESCE" }
+};
+
+QString QgsHanaExpressionCompiler::sqlFunctionFromFunctionName( const QString &fnName ) const
+{
+  return FUNCTION_NAMES_SQL_FUNCTIONS_MAP.value( fnName, QString() );
+}
+
+QString QgsHanaExpressionCompiler::castToReal( const QString &value ) const
+{
+  return QStringLiteral( "CAST((%1) AS REAL)" ).arg( value );
+}
+
+QString QgsHanaExpressionCompiler::castToInt( const QString &value ) const
+{
+  return QStringLiteral( "CAST((%1) AS INTEGER)" ).arg( value );
+}
+
+QString QgsHanaExpressionCompiler::castToText( const QString &value ) const
+{
+  return QStringLiteral( "CAST((%1) AS NVARCHAR)" ).arg( value );
+}
+
+QgsSqlExpressionCompiler::Result QgsHanaExpressionCompiler::compileNode(
+  const QgsExpressionNode *node, QString &result )
+{
+  switch ( node->nodeType() )
+  {
+    case QgsExpressionNode::ntFunction:
+    {
+      const QgsExpressionNodeFunction *nodeFunc = static_cast<const QgsExpressionNodeFunction *>( node );
+      QgsExpressionFunction *fd = QgsExpression::Functions()[nodeFunc->fnIndex()];
+
+      QString funcName = fd->name();
+      if ( funcName.isEmpty() )
+        break;
+
+      if ( funcName == QLatin1String( "$geometry" ) )
+      {
+        result = quotedIdentifier( mGeometryColumn );
+        return Complete;
+      }
+      else if ( funcName.toLower() == QLatin1String( "pi" ) )
+      {
+        result = "3.141592653589793238";
+        return Complete;
+      }
+    }
+    break;
+    case QgsExpressionNode::ntLiteral:
+    {
+      const QgsExpressionNodeLiteral *n = static_cast<const QgsExpressionNodeLiteral *>( node );
+      switch ( n->value().type() )
+      {
+        case QVariant::Bool:
+          result = n->value().toBool() ? "(1=1)" : "(1=0)";
+          return Complete;
+        default:
+          break;
+      }
+    }
+    break;
+    case QgsExpressionNode::ntUnaryOperator:
+    {
+      const QgsExpressionNodeUnaryOperator *unaryOp = static_cast<const QgsExpressionNodeUnaryOperator *>( node );
+      switch ( unaryOp->op() )
+      {
+        case QgsExpressionNodeUnaryOperator::uoNot:
+        {
+          Result resRight = compileNode( unaryOp->operand(), result );
+          if ( "NULL" == result.toUpper() )
```

Comment:
```suggestion
          if ( QLatin1String( "NULL" ) == result.toUpper() )
```

@mrylov can you do another iteration, checking for any raw literals.

Reply:
Done in 9925cabf.